### PR TITLE
feat(retry): Add support for backoff strategies and jitter

### PR DESCRIPTION
**Pull request by rjvitorino**
This pull request adds support for different backoff strategies (sequential, logarithmic, exponential) and optional jitter to the `@retry` decorator, addressing the need for more flexible retry mechanisms as described in issue #3.

### Summary of Changes
- **Backoff Strategies**: Implemented sequential, logarithmic, and exponential backoff strategies.
- **Jitter**: Added optional jitter to backoff intervals to help prevent the thundering herd problem.
- **Documentation**: Updated the README.md to document these new features.
- **Tests**: Added unit tests for the new backoff strategies and jitter functionality.

### Backoff Strategies
The `@retry` decorator now supports the following backoff strategies:
- **Sequential**: Linear backoff, incrementing the wait time by a fixed amount each retry.
- **Logarithmic**: Logarithmic backoff, using the natural logarithm to increment the wait time.
- **Exponential**: Exponential backoff, doubling the wait time each retry.

### Jitter
An optional `jitter` parameter has been added to introduce a random amount of time to each backoff interval, helping to mitigate the thundering herd problem.

### Usage
Here is an example of how to use the updated `@retry` decorator:

```python
import refry

@refry.retry(rate_limit_exception=Exception, backoff_increment=2, retries=5, backoff_type="sequential", jitter=True)
def function_with_problem():
    raise Exception('Something bad happens here')

function_with_problem()
```

### Tests
Unit tests have been added to cover:

* Sequential backoff strategy
* Logarithmic backoff strategy
* Exponential backoff strategy
* Sequential backoff with jitter
* Logarithmic backoff with jitter
* Exponential backoff with jitter

These changes improve the flexibility and robustness of the `@refry.retry` decorator, making it suitable for a wider range of use cases.

### Issue Fixed
This pull request fixes issue #3 (except the multiple types of server errors/exceptions)

### Checklist
* [x] Added tests for new features
* [x] Updated documentation
* [x] Verified that all tests pass

Thank you for reviewing this pull request. @pydanny I look forward to your feedback.

### CODE REVIEW

**Comment by rjvitorino** [comment]
@pydanny did you check this?

File: src/refry/main.py [hunk 1]
```diff
@@ -1,30 +1,46 @@
...
     """
 
+    def calculate_backoff(attempt: int) -> float:
+        backoff_strategies = {
+            "sequential": backoff_increment * (attempt + 1),
+            "logarithmic": math.log(attempt + 2) * backoff_increment,
+            "exponential": (2**attempt) * backoff_increment,
+        }
+        backoff = backoff_strategies.get(
+            backoff_type, backoff_increment * (attempt + 1)
```
Comment (pydanny) [inline src/refry/main.py:28]:
```suggestion
            # Default is the sequential backoff strategy
            backoff_type, backoff_strategies['sequential']
```

File: src/refry/main.py [hunk 2]
```diff
@@ -1,30 +1,46 @@
...
+        )
+        return backoff + random.uniform(0, backoff_increment) if jitter else backoff
```
Comment (pydanny) [inline src/refry/main.py:30]: Until recently I didn't like ternary operators in Python, wishing for the C-style method. However, I'm starting to come around. No change necessary. 😄

**Review by pydanny — CHANGES_REQUESTED** [review]
Excellent work, concise yet clear in implementation. 👍🏻 

I've got a small request, and made a no-change needed comment. Otherwise it is good to go. I'm on a coding holiday so when I'll check on changes you may have made daily.

File: src/refry/main.py (hunk 2 above)
Comment (rjvitorino) [reply src/refry/main.py:31]: Thank you kindly @pydanny! It's great to read your feedback; I am thrilled to help 😄 

I integrated your comment (good point!), and I'm glad you're starting to come around ternary operators 😅 I like them quite much!

**Review by pydanny — APPROVED** [review]
Awesome!

**Pull request merged**